### PR TITLE
Filter an array containing the file to be updated

### PR DESCRIPTION
**Pull request by MarkGibbons**
Work in progress to show how I think using a ruby proc as a filter to do arbitrary things with a file might work.  Related to #95 and #79.  Took some of the ideas from poise-line.

Include filters for inserting lines before or after a match.
Introduces a new resource filter_lines. 

Attributes:
property :filter, [Method, Proc] - a proc or a method.  I've put some default methods in Line::Filter
property :filter_args, Array - arguments specific to the filter, might be lines to insert, regex, anything really
property :path, String - file to change

For example the :before filter can be used to insert a bunch of lines immediately before a matching regex.  Arguments are pattern, array of lines to insert, match and insert (each, first only, last)

The tests show some inline procs create on the fly.

I don't really like sending the arguments as positional parms in an array.  I'm not sure about using a method instead of a proc.  It works but is there something in ruby that makes it a bad idea.

The :before and :after filters need some love. They are first working draft code and are a bit too much like fortran instead of ruby.  People could write procs and extend the Line::Filter class or use their own filter class.

Thanks,
  Mark

---
This change is [Reviewable](https://reviewable.io/reviews/sous-chefs/line-cookbook/101)

### CODE REVIEW

**Comment by kitchen-porter** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big Pull Request.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: test/fixtures/cookbooks/test/recipes/filter_lines.rb [hunk 1]
```diff
@@ -0,0 +1,125 @@
+directory '/tmp'
+
+template '/tmp/dangerfile' do
+end
+
```
Comment (MarkGibbons) [inline test/fixtures/cookbooks/test/recipes/filter_lines.rb]: Some examples of what's possible.  Inspec tests to check the results are there too.

File: test/fixtures/cookbooks/test/recipes/filter_lines.rb [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+  filter filters.method(:before)
+  filter_args [match_pattern, insert_lines]
+end
+
+template '/tmp/before_first' do
+  source 'dangerfile.erb'
+  sensitive true
+end
+
+filter_lines 'Insert lines before match' do
```
Comment (MarkGibbons) [inline test/fixtures/cookbooks/test/recipes/filter_lines.rb:84]: Insert one time example.  Insert before the first match.

File: test/fixtures/cookbooks/test/recipes/filter_lines.rb [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+  path '/tmp/before'
+  sensitive false
+  filter filters.method(:before)
+  filter_args [match_pattern, insert_lines]
+end
+
+filter_lines 'Insert lines before match 2nd try' do
```
Comment (MarkGibbons) [inline test/fixtures/cookbooks/test/recipes/filter_lines.rb]: Idempotence check

File: test/fixtures/cookbooks/test/recipes/filter_lines.rb [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+insert_lines = %w(line1 line2 line3)
+match_pattern = /^COMMENT ME|^HELLO/
+
+# ==================== before filter =================
+template '/tmp/before' do
+  source 'dangerfile.erb'
+  sensitive true
+end
+
+filter_lines 'Insert lines before match' do
```
Comment (MarkGibbons) [inline test/fixtures/cookbooks/test/recipes/filter_lines.rb:84]: Insert multiple times.  Before each match of the regex.

File: test/fixtures/cookbooks/test/recipes/filter_lines.rb [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+filter_lines 'Do nothing' do
+  path '/tmp/dangerfile'
+  filter proc { |current| current }
+end
+
+template '/tmp/reverse' do
+  source 'dangerfile.erb'
+end
+
+filter_lines 'Reverse line text' do
```
Comment (MarkGibbons) [inline test/fixtures/cookbooks/test/recipes/filter_lines.rb:18]: Arbitrary in line proc.  Can do anything with the array as long as a new array is returned.

**Comment by MarkGibbons** [comment]
Review request sent to Lamont to see if this code is generally the direction he meant by just run ruby against the file.

thanks

**Comment by lamont-granquist** [comment]
Well I've got a patch I'm working on against core, right now it will do things like:

```ruby
file "/tmp/one" do
  source "one"
  edit do
    empty!
    delete_lines_matching /FOO/
    append_if_no_such_line "FOOBAR"
  end
end
```

So it takes a block, but has DSL entries that correspond to operations.   That makes all the actions of this cookbook composable in one resource -- and on any file resource.

I think what you're working on here could probably be plugged into that DSL as a "filter" or "apply_proc" or something...

**Comment by damacus** [comment]
I was taking to Tim about this yesterday. It might be worth throwing out the api to users using this cookbook then porting those to core if people find it useful

**Comment by lamont-granquist** [comment]
That would mean a monkeypatch to the core file resource in a cookbook though.

**Comment by lamont-granquist** [comment]
(plus that would mean we could not remove the core usage of FileEdit since we can't have core resources depend on a monkeypatch in a cookbook)

**Comment by MarkGibbons** [comment]
Unless it's just a total waste of time I'll keep working on the filter_lines resource.  I can create filters for things people of asked for (mostly around inserting or changing lines in stanzas).  I have been looking at augeas for a couple years and have a few uses.  Multiple changes in one resource run is also something I might take a run at for my own amusement.

I don't understand "That would mean a monkeypatch to the core file resource" or how that would involve FileEdit.  Could anyone explain the problem a bit more?   I think you might mean hooking the filter_lists code into the file resource via a monkey patch??

**Comment by lamont-granquist** [comment]
yeah, i'm still talking about my patch...

this still looks interesting...

File: libraries/filter.rb [hunk 6]
```diff
@@ -2,11 +2,13 @@
 module Line
   class Filter
     def before(current, args)
-      # Insert a set of lines before each match of the pattern
+      # Insert a set of lines immediately before each match of the pattern - seems like this has limited usefulness.  Comments and blank like would screw it up.
```
Comment (damacus) [inline libraries/filter.rb]: there's a wee typo here

Comment (MarkGibbons) [reply libraries/filter.rb]: The before and after filters as they stand are unlikely to make the final cut.  I'm trying to figure out something more useful.

**Comment by MarkGibbons** [comment]
Example of what can be done.  This resource will insert some lines before a match.  Insert some lines after a match and the delete the comments from the file.  The user interface is a really nasty array of hash entries with a nested array.

filters = Line::Filter.new  # Use the built in filters
insert_lines = %w(line1 line2 line3)  # test lines to insert
match_pattern = /^COMMENT ME|^HELLO/  # line to insert around

````
filter_lines 'Multiple before and after match' do
  path '/tmp/multiple_filters'
    filters(
      [
        # insert lines before the last match
        { code: filters.method(:before),          
          args: [match_pattern, insert_lines, :last]
        },
        # insert lines after the last match
        { code: filters.method(:after),
          args: [match_pattern, insert_lines, :last]
        },
       # delete comment lines
       { code: proc { |current| current.select do |line| line !~ /^#/ end }
        }
      ]
   )
 end
````

**Comment by MarkGibbons** [comment]
I've added 5 filters, a method to use one of the supplied filters or a user supplied proc and a method to use multiple filters.  That's about it for me, all the edit requests I know of have been covered.  I need to do some polishing and handle some error conditions, but the essential functions are there.

**Comment by MarkGibbons** [comment]
Removed WIP.  I could use some other eyes pointing out problems or suggestions for improvements.

**Comment by MarkGibbons** [comment]
Back to WIP status.  I'm going to redo the resource interface.  I think it can be a lot simpler.

**Comment by MarkGibbons** [comment]
Redid the user interface.  I think it's much easier to use and understand.  Ready for a review.

**Review by damacus — COMMENTED** [review]
> Reviewed 3 of 31 files at r7, 1 of 8 files at r10, 1 of 5 files at r12.
> *[Reviewable](https://reviewable.io/reviews/sous-chefs/line-cookbook/101)* status: 5 of 47 files reviewed, all discussions resolved (waiting on @MarkGibbons)

**Pull request merged**